api/controllers/AssignmentController.js

```diff
@@ -119,14 +98,60 @@ module.exports = class AssignmentController extends classMixin(OrderedDataContro
       )
     }
 
-    if (judgement.grade > 0 && !parameterizedAssignment.dataValues.solved) {
-      await parameterizedAssignment.update({ solved: true })
-    }
-
     return res.status(200).send(
       {
         success: true, judgement
       }
     )
   }
+
+  async processAsyncJudge (req, res, aux, assignment, solution, parameterizedAssignment) {
+    let judgement = { grade: 0, msg: 'Wrong! Please try again.' }
+    const judge = new assignment.Judge(assignment.judge, req.user)
+
+    res.status(202).send({ code: 202, judgement: { grade: 0, msg: 'Solution is being processed. Please wait!' } })
+
+    try {
+      judgement = await judge.judge(aux, req.user, assignment.Judge, solution)
+      await this.updateSolution(req, parameterizedAssignment, solution)
+      await this.updateSolved(judgement, parameterizedAssignment)
+      appEmitter.emit('solution', { ...judgement })
+    } catch (err) {
+      logger.error(err)
+      judgement = { grade: 0, msg: err.message }
+
+      appEmitter.emit('solution', { ...judgement })
+    }
```

Comment:
DRY: `processAsyncJudge` and `processSyncJudge` are almost identical. Abstract out this behavior into a `.evaluateSolution` function which accepts two callback functions: `correct(judgement)` and `incorrect(judgement)`. Then `asyncJudge` and `syncJudge` can specialize by providing different callbacks.